splight_cli/solution/state.py

```diff
@@ -0,0 +1,162 @@
+
+
+class ResourceAlreadyExistsError(Exception):
+    def __init__(self, id):
+        super().__init__(f"Resource with ID '{id}' already exists.")
+
+
+class FileHandler(BaseModel):
+    @abstractmethod
+    def load(self) -> Dict:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def save(self) -> None:
+        raise NotImplementedError()
+
+
+class LocalFileHandler(FileHandler):
+    path: FilePath
+
+    def load(self) -> Dict:
+        with open(self.path, "r") as fp:
+            return json.load(fp)
+
+    def save(self, state_data: Dict) -> None:
+        with open(self.path, "w") as fp:
+            json.dump(state_data, fp)
+
+
+class S3FileHandler(FileHandler):
+    bucket: str
+    key: str
+
+    _settings: BaseSettings = PrivateAttr()
+    _client = PrivateAttr()
+
+    @property
+    def _settings(self):
+        return AWSSettings()
+
+    @property
+    def _client(self):
+        client = boto3.client(
+            "s3",
+            region_name=self._settings.DEFAULT_REGION,
+            aws_access_key_id=self._settings.ACCESS_KEY_ID,
+            aws_secret_access_key=self._settings.SECRET_ACCESS_KEY,
+        )
+        return client
+
+    def load(self) -> Dict:
+        response = self._client.get_object(Bucket=self.bucket, Key=self.key)
+        state_data = json.loads(response["Body"].read().decode("utf-8"))
+        return state_data
+
+    def save(self, state_data: Dict) -> None:
+        self._client.put_object(
+            Bucket=self.bucket,
+            Key=self.key,
+            Body=json.dumps(state_data),
+            ContentType="application/json",
+        )
+
+
+class StateData(BaseModel):
+    """
+    Represents the state data with Pydantic properties for serialization.
+
+    # NOTE: Pydantic properties declared here will be included in the statefile.
+
+    Attributes:
+        resource_map (Dict): A dictionary to store resource data.
+
+    Methods:
+        get(id: str) -> Dict:
+            Retrieve resource data based on the provided ID.
+
+        add(id: str, data: Dict) -> None:
+            Add new resource data with the specified ID.
+
+        delete(id: str) -> None:
+            Delete resource data associated with the given ID.
+
+        update(id: str, data: Dict) -> None:
+            Update resource data for the specified ID.
+
+    Raises:
+        UnexistingResourceException: Raised when attempting to access or update
+                                     a non-existing resource.
+        ResourceAlreadyExistsError: Raised when trying to add a resource with
+                                    an ID that already exists in the resource map.
+    """
+
+    resource_map: Dict = {}
+
+    def get(self, id: str) -> Dict:
+        if id not in self.resource_map:
+            raise UnexistingResourceException(id)
+        return self.resource_map[id]
+
+    def add(self, id: str, data: Dict) -> None:
+        if not id in self.resource_map:
```

Comment:
Although this is the same as doing `if id not in self.resource_map` I would stick to one way of writing it to be consistent